File: src/test/java/com/koddy/server/member/domain/model/mentor/MentoringPeriodTest.java
package com.koddy.server.member.domain.model.mentor;

import com.koddy.server.common.UnitTest;
import com.koddy.server.member.exception.MemberException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.time.LocalDateTime;

import static com.koddy.server.member.domain.model.mentor.MentoringPeriod.TimeUnit.HALF_HOUR;
import static com.koddy.server.member.exception.MemberExceptionCode.SCHEDULE_PERIOD_TIME_MUST_ALIGN;
import static com.koddy.server.member.exception.MemberExceptionCode.SCHEDULE_PERIOD_TIME_MUST_EXISTS;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;

@DisplayName("Member/Mentor -> 도메인 [MentoringPeriod] 테스트")
class MentoringPeriodTest extends UnitTest {
    @Nested
    @DisplayName("MentoringPeriod 생성")
    class Construct {
        @Test
        @DisplayName("날짜가 입력되지 않으면 MentoringPeriod 생성에 실패한다")
        void throwExceptionByDateNotExists() {
            assertThatThrownBy(() -> MentoringPeriod.of(LocalDate.of(2024, 1, 1), null))
                    .isInstanceOf(MemberException.class)
                    .hasMessage(SCHEDULE_PERIOD_TIME_MUST_EXISTS.getMessage());
        }

        @Test
        @DisplayName("멘토링 종료날짜가 시작날짜 이전이면 MentoringPeriod 생성에 실패한다")
        void throwExceptionByStartEndNotAlign() {
            final LocalDate startDate = LocalDate.of(2024, 3, 1);
            final LocalDate endDate = LocalDate.of(2024, 2, 1);

            assertThatThrownBy(() -> MentoringPeriod.of(startDate, endDate))
                    .isInstanceOf(MemberException.class)
                    .hasMessage(SCHEDULE_PERIOD_TIME_MUST_ALIGN.getMessage());
        }

        @Test
        @DisplayName("MentoringPeriod을 생성한다")
        void construct() {
            // given
            final LocalDate startDate = LocalDate.of(2024, 2, 1);
            final LocalDate endDate = LocalDate.of(2024, 3, 1);

            // when
            final MentoringPeriod mentoringPeriod = MentoringPeriod.of(startDate, endDate);

            // then
            assertAll(
                    () -> assertThat(mentoringPeriod.getStartDate()).isEqualTo(startDate),
                    () -> assertThat(mentoringPeriod.getEndDate()).isEqualTo(endDate),
                    () -> assertThat(mentoringPeriod.getTimeUnit()).isEqualTo(HALF_HOUR)
            );
        }
    }

    @Test
    @DisplayName("주어진 날짜가 멘토링 진행 기간에 포함되는지 확인한다")
    void isDateIncluded() {
        // given
        final LocalDate startDate = LocalDate.of(2024, 2, 1);
        final LocalDate endDate = LocalDate.of(2024, 3, 1);
        final MentoringPeriod mentoringPeriod = MentoringPeriod.of(startDate, endDate);

        // when
        final boolean actual1 = mentoringPeriod.isDateIncluded(LocalDate.of(2024, 2, 1));
        final boolean actual2 = mentoringPeriod.isDateIncluded(LocalDate.of(2024, 2, 15));
        final boolean actual3 = mentoringPeriod.isDateIncluded(LocalDate.of(2024, 3, 1));
        final boolean actual4 = mentoringPeriod.isDateIncluded(LocalDate.of(2024, 3, 15));

        // then
        assertAll(
                () -> assertThat(actual1).isTrue(),
                () -> assertThat(actual2).isTrue(),
                () -> assertThat(actual3).isTrue(),
                () -> assertThat(actual4).isFalse()
        );
    }

    @Test
    @DisplayName("멘토링 진행 시간이 멘토의 TimeUnit이랑 일치하는지 확인한다")
    void allowedTimeUnit() {
        // given
        final LocalDate startDate = LocalDate.of(2024, 2, 1);
        final LocalDate endDate = LocalDate.of(2024, 3, 1);
        final MentoringPeriod mentoringPeriod = MentoringPeriod.of(startDate, endDate); // TODO default = 30

        // when
        final LocalDateTime standard = LocalDateTime.of(2024, 2, 15, 18, 0);

        final boolean actual1 = mentoringPeriod.allowedTimeUnit(standard, standard.plusMinutes(30));
        final boolean actual2 = mentoringPeriod.allowedTimeUnit(standard, standard.plusMinutes(60));

        // then
        assertAll(
                () -> assertThat(actual1).isTrue(),
                () -> assertThat(actual2).isFalse()
        );
    }
}

File: src/test/kotlin/com/koddy/server/member/domain/model/mentor/MentoringPeriodTest.kt
package com.koddy.server.member.domain.model.mentor

import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.toLocalDate
import com.koddy.server.common.toLocalDateTime
import com.koddy.server.member.exception.MemberException
import com.koddy.server.member.exception.MemberExceptionCode.SCHEDULE_PERIOD_TIME_MUST_ALIGN
import io.kotest.assertions.assertSoftly
import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.FeatureSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.throwable.shouldHaveMessage
import java.time.LocalDateTime

@UnitTestKt
@DisplayName("Member/Mentor -> 도메인 [MentoringPeriod] 테스트")
internal class MentoringPeriodTest : FeatureSpec({
    feature("MentoringPeriod 생성") {
        scenario("Start가 End보다 나중 날짜로 설정되면 MentoringPeriod을 생성할 수 없다") {
            shouldThrow<MemberException> {
                MentoringPeriod(
                    startDate = "2024/3/2".toLocalDate(),
                    endDate = "2024/3/1".toLocalDate(),
                )
            } shouldHaveMessage SCHEDULE_PERIOD_TIME_MUST_ALIGN.message
        }

        scenario("Start가 End보다 같거나 느린 날짜로 설정되면 MentoringPeriod을 생성할 수 있다") {
            val table = mapOf(
                "2024/3/1".toLocalDate() to "2024/3/1".toLocalDate(),
                "2024/3/1".toLocalDate() to "2024/3/2".toLocalDate(),
            )

            table.forEach { (startDate, endDate) ->
                shouldNotThrowAny { MentoringPeriod(startDate = startDate, endDate = endDate) }
            }
        }
    }

    feature("MentoringPeriod's isDateIncluded") {
        scenario("주어진 날짜가 멘토링 진행 기간에 포함되는지 확인한다") {
            val mentoringPeriod = MentoringPeriod(
                startDate = "2024/3/1".toLocalDate(),
                endDate = "2024/3/10".toLocalDate(),
            )

            mapOf(
                "2024/2/26".toLocalDate() to false,
                "2024/3/1".toLocalDate() to true,
                "2024/3/4".toLocalDate() to true,
                "2024/3/10".toLocalDate() to true,
                "2024/3/11".toLocalDate() to false,
            ).forEach { (target, result) ->
                mentoringPeriod.isDateIncluded(target) shouldBe result
            }
        }
    }

    feature("MentoringPeriod's allowedTimeUnit") {
        scenario("멘토링 진행 시간이 멘토의 TimeUnit이랑 일치하는지 확인한다") {
            val timeUnit30 = MentoringPeriod(
                startDate = "2024/3/1".toLocalDate(),
                endDate = "2024/3/10".toLocalDate(),
            )
            val timeUnit60 = MentoringPeriod(
                startDate = "2024/3/1".toLocalDate(),
                endDate = "2024/3/10".toLocalDate(),
                timeUnit = MentoringPeriod.TimeUnit.ONE_HOUR,
            )

            val start: LocalDateTime = "2024/3/1-18:00".toLocalDateTime()
            assertSoftly {
                timeUnit30.allowedTimeUnit(start, start.plusMinutes(30)) shouldBe true
                timeUnit30.allowedTimeUnit(start, start.plusMinutes(60)) shouldBe false
                timeUnit60.allowedTimeUnit(start, start.plusMinutes(30)) shouldBe false
                timeUnit60.allowedTimeUnit(start, start.plusMinutes(60)) shouldBe true
            }
        }
    }
})
